\documentclass[12pt]{article}
\usepackage{amsmath}
\usepackage{bm}
\usepackage[colorlinks=true,urlcolor=blue]{hyperref}

\title{Snp HeRitability Estimation Kit (SHREK) Development Log}
\date{\today}
\author{Sam Shing Wan, Choi\\ 
	\texttt{choishingwan@gmail.com}
	\and
	Johnny Sheung Him, Kwan\\
	\texttt{shkwan@hku.hk}
	\and
	Henry Chi-Ming, Leung\\
	\texttt{cmleung2@cs.hku.hk}
}


\begin{document}
	
	\maketitle
	\tableofcontents
	
	\section{Assumption in implementation}
	To make life easier, I made a number of assumption when I implement the programme and they are listed as follow
	\begin{enumerate}
		\item The intervals within each individual bed files are non-overlapping
		\item SNPs filtered out when reading the p-value file will not be included in the output no matter what (avoid storing unnecessary information)
		\item We will remove \emph{all} SNPs that are in perfect LD with each-other
		\begin{enumerate}
			\item Realistically, we can not remove \emph{all} SNPs that are in perfect LD with each-other without first computing the LD matrix for the whole genome. 
			However, that will incur a large speed penalty to the algorithm.
			To compromise, we remove SNPs that are in perfect LD with each-other \emph{within the same window}.
			Take into consideration of the calculation of variance, we will need to plan forward and consider SNPs within 4 windows. 
			That's definitely overkill but that avoid problem of crazy perfect LD interactions (to be honest, I would consider those as error, yet it is easier for me to just remove them)
		\end{enumerate}
		\item We do not consider a SNP ambiguous unless we need to perform flipping
	\end{enumerate}
	
	\section{January 25, 2016}
	We are going to re-write SHREK starting from today. 
	There are a number of main goal in this re-work
	\begin{enumerate}
		\item The whole parameter parsing 
		\item Better help messages
		\item Allow filtering by imputation score (0-1)
		\item Filtering of SNPs that have different reference/alternative allele with reference
		\item Use Armadillo instead of EIGEN to improve speed
		\item Filter \emph{all} SNPs that are in perfect LD
		\item Implement the advance variance calculation 
		\item Start documenting the codes
	\end{enumerate}
	\subsection{Finished today}
	\begin{enumerate}
		\item Finished the error message for different modes
		\item Finished the parsing for binary trait
		\item Finished the parsing for quantitative trait
		\item Disabled risk prediction, should focus on the heritability estimation first
	\end{enumerate}
	\emph{I have not perform debugging, there can be error}
	\section{January 26, 2016}
	Continue on where I left yesterday
	Assumptions we made in the programme
	\begin{enumerate}
		\item p-value file has header
		\item the numbers return from Command class are index (0 based)
	\end{enumerate}
	I really want to write a more compacted code.
	Will try to restructure the command class
	\subsection{Finished today}
	\begin{enumerate}
		\item Finished refining the usage messages
		\item Finished the basic parameter parsing (Still haven't finish the index based parameter though)
	\end{enumerate}
	\section{January 27, 2016}
	\subsection{Aim}
	We want to at least finish the region parsing and SNP class update today
	\subsection{Finished today}
	\begin{enumerate}
		\item Completed the update of the Command class (compiled without problem)
		\item Completed the update of the Region class (compiled without problem)
		\item Completed the update of the Snp class (compiled without problem)
	\end{enumerate}
	I also \emph{haven't} test run the programme, so there might be additional syntax / logic error that are not picked up at the current stage.
	
	\section{January 28, 2016}
	\subsection{Finished today}
	\begin{enumerate}
		\item Finished the checking of reference panel
		\item Found a different algorithm for popcount, which should be platform independent
		\item Installed armadillo and OpenBLAS
		\begin{enumerate}
			\item I have to add the OpenBLAS to the PATH and LD\_LIBRARY\_PATH for armadillo to detect it
		\end{enumerate}
		\item Updated the makefile accordingly
		\begin{enumerate}
			\item However, I have not updated the decomposition class, therefore there will be problem when we try to compile the programme
		\end{enumerate}
		\item Updated the $r^2$ calculation function such that it is using the correct sample size
		\begin{enumerate}
			\item The sample size when calculating the $r^2$ is defined as the number of samples containing \emph{both} SNPs
			\item Because of how complicated the genotype class is, I really don't want to modify it
			\item The main bottleneck for the computation is in the decomposition anyway, so it should be fine
		\end{enumerate}
		
	\end{enumerate}
	
	\section{January 29, 2016}
	Need to at least finish the getSNP function.
	Must be careful with it as this is not only complicated, but also extremely important for the whole programme.
	
	I spent whole day in annotating the document of the class to make sure I don't overlook any problem.
	
	\section{February 1, 2016}
	Start to implement the getSNP function. 
	A number of point to note
	\begin{enumerate}
		\item First get all the SNPs that passed all filtering (e.g. MAF)
		\item Then for each window, calculate the LD
		\item When encountered with perfect LD, we retain the \emph{first} SNP
		\begin{enumerate}
			\item The summary statistics will be the mean of all SNPs in perfect LD
			\item All SNPs except the first are \emph{removed} 
			\item So they will not appear in subsequent analysis
		\end{enumerate}
	\end{enumerate}
	
	\section{February 2, 2016}
	Try to perform extensive check on codes that has been written
	\subsection{Finished today}
	\begin{enumerate}
		\item Finished debugging, everything up until the initialization of genotypeFileHandler is ok
	\end{enumerate}
	\section{February 3, 2016}
	Sat in a meeting with Pak today, now we have a few things to-do.
	\subsection{To Do}
	\begin{enumerate}
		\item Compare the condition number of the LD matrix before and after the LD correction
		\item Compare the performance of case control estimation with different sample size 
		\item For LD correction, might also worth testing how the density and total number of SNPs affect the estimation
		\item For effect size estimation, don't use the complicated method
	\end{enumerate}
	\subsection{Finished today}
	We have switched to use list instead of deque. 
	The benefit of list is that it is a double linked list, therefore it is much easier for one to remove the middle part.
	Such property is ideal for the handling of perfect LD. 
	However, it also means that we cannot access the elements of the list by index. 
	The only way for us to access the elements is to use the iterators and the advance function from the iterator library.
	Although this makes the implementation more difficult, it should be able to help us to speed up our code even if we want to perform perfect LD removal.
	\begin{enumerate}
		\item Extensive testing of the previous classes
		\item Now use iterators for the boundary, this is more dynamic and can work with the perfect LD removal without inducing much complication into the algorithm (however, the implementation itself can be complicated)
	\end{enumerate}
	\subsection{Problem identified today}
	So assuming that the SNPs A B C are occurring in the genome by alphabetical order, when we calculate the LD of A with B C, we do not observe any perfect LD. 
	This will lead us to continue with the calculation of SNP B and C. 
	However, if SNP B and SNP C is in perfect LD, we will need to go back to the LDs of SNP A to remove the column of SNP C.
	Not only is this procedure complicates the implementation, it will also significantly slow down the computation because of the constant resizing of the matrix.
	
	To conclude, it might be better for us to use the original ``sausage'' approach, where we should also locate all the perfect LD location.
	Then we can remove the perfect LD afterwards.
	
	\section{February 4, 2016}
	Was spending the whole day in implementing the ``sausage'' approach.
	However, it was found that the calculated $R^2$ are terribly wrong. 
	So now converting back to the old method of inclusion for the genotype reading.
	We will then check the implementation of the $R^2$ computation right after we have make sure the genotype reading is correct.
	\section{February 5, 2016}
	Try to continue on what we have been doing yesterday.
	
\end{document}